### PR TITLE
Supergreedy Model Refactor + merged with master

### PR DESCRIPTION
**Pull request by alexcuozzo-1**
https://ntaresearch.atlassian.net/browse/RES-2541

### CODE REVIEW

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/experiments/block_model_experiment.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    nn.ReLU()
+    BilinearInfo()
+
+    This can get to be very complicated for things like large ResNets, so the default
+    set of ResNets with various sparsity levels have been pre-defined in
+    greedy_infomax/utils/model_utils.py.
+    """
+    def setup_experiment(self, config):
+        if config.get("cuda_launch_blocking", False):
+            os.environ["CUDA_LAUNCH_BLOCKING"] = "1"
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/experiments/block_model_experiment.py:99]: It turns out that this is too late to apply this command (and should be done in the `run.py` file instead). You should just remove this option since it's purely for debugging, and not something we want to make public.

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/bilinear_info.py [hunk 2]
```diff
@@ -211,3 +208,133 @@ def __init__(
             )
             for _ in range(self.k_predictions)
         )
+
+
+class _BilinearInfo(nn.Module):
+    """
+    From the Greedy InfoMax paper, a module which estimates the mutual information
+    between two representations through a bilinear model. This is now named
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/bilinear_info.py:216]: This docstring should look something like the following to make it clear to the user that it's deprecated:

```
class _BilinearInfo(nn.Module):
    """
    NOTE: This class is deprecated.

    ...
    """
```

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/block_model.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    contains a list of modules which are sequentially iterated through for the
+    forward/encode pass. The distinction between supervised training and unsupervised
+    training is implemented by switching which modules are used for the forward pass
+    versus the encode pass. In the forward pass, the BilinearInfoModules are active
+    and the EmitEncoding module is inactive. In the encode pass, the BilinearInfo
+    modules are inactive and the EmitEncoding modules are active.
+    """
+    def __init__(self, modules, **kwargs):
+        super(BlockModel, self).__init__()
+        self.module_list = nn.ModuleList(modules)
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/block_model.py:51]: It would be super helpful to include the params for the `__init__` method in the docstring above, just as you did for `FlattenClassifier`.

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/classification_model.py [hunk 4]
```diff
@@ -48,12 +48,47 @@ def __init__(self, in_channels=256, num_classes=10):
...
+        x = x.detach()
+        x = self.avg_pool(x).squeeze()
+        x = self.model(x).squeeze()
+        return x.view(batch_size, -1)
+
+
+class FlattenClassifier(nn.Module):
+    """
+    A simple multilayer perceptron classification head which outputs a distribution
+    over possible class labels. This is used in the supervised phase of Greedy
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/classification_model.py:64]: It's not clear from the docstring that you're taking _z_'s corresponding to different parts of an image and flattening them into 1 long vector; my initial thought was you're taking the output of a conv layer and removing the _x_ or _y_ dimension. Can you please make this clear?

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/full_model.py [hunk 5]
```diff
@@ -52,8 +52,14 @@ class FullVisionModel(torch.nn.Module):
     A modified version of ResNet to compute patch-wise representations. This model
     is the encoder in self-supervised experiments and does not include a built in
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/full_model.py:53]: Are you still using these "full models" despite the more recent addition of specifying block models instead? It's important to know, since if you're not, then we can move this file to the archive folder.

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        out = out.permute(0, 3, 1, 2).contiguous()
+        info_estimate = self.bilinear_info.estimate_info(out, out)
+        self.info_estimate_aggregator.append(info_estimate)
+        x_blocked = self.gradient_block(x).clone()
+        return x_blocked
+
+    """
+    During unsupervised training, this function will be linked to the forward hook
+    for its corresponding module.
+    """
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py:124]: It's odd to have """ doctrings outside of a function, can you replace this with regular comments (`#`)?

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    A version of the above GreedyInfoMaxBlock which uses SparseBilinearInfo instead
+    of a regular BilinearInfo module.
+    """
+    def __init__(self,
+                 estimator_outputs,
+                 encoding_outputs,
+                 in_channels,
+                 negative_samples=16,
+                 k_predictions=5,
+                 sparsity=None):
```
Comment (karangrewal) [inline packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py:151]: `sparsity` shouldn't be None, it should default to some numeric value instead.

File: packages/greedy_infomax/tests/unit/test_gradient_block.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+from torch.utils.data import DataLoader
+from torchvision.datasets import FakeData
+from torchvision.transforms import ToTensor
+
+from nupic.research.frameworks.greedy_infomax.models.utility_layers import GradientBlock
+
+
+class TestGradientBlock(unittest.TestCase):
+
+    def test_gradients_blocked(self):
```
Comment (karangrewal) [inline packages/greedy_infomax/tests/unit/test_gradient_block.py:36]: I recommend doing the following, which should't take more than 30 secs:

Add another test case that's identical to this one, except that you don't have a `GradientBlock` module in `model`, and test to see if the grads aren't None. That will be more convincing that your gradient block is working.

File: projects/greedy_infomax/experiments/__init__.py [hunk 9]
```diff
@@ -23,9 +23,19 @@
 from .sigopt_experiments import CONFIGS as SIGOPT_EXPERIMENTS
 from .small_sparse import CONFIGS as SMALL_SPARSE
 from .sparse_resnets import CONFIGS as SPARSE_RESNETS
+from .block_wise_training import CONFIGS as BLOCK_WISE
```
Comment (karangrewal) [inline projects/greedy_infomax/experiments/__init__.py:26]: (low priority) Run `isort` on this file to alphabetize imports.

**Review by karangrewal — CHANGES_REQUESTED** [review]
Here's my feedback, and I may add more comments later.

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/experiments/block_model_experiment.py (hunk 1 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/experiments/block_model_experiment.py:99]: Done

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/bilinear_info.py (hunk 2 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/bilinear_info.py:216]: Makes sense, fixed

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/block_model.py (hunk 3 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/block_model.py:51]: Done

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/classification_model.py (hunk 4 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/classification_model.py:64]: Yes, I'll explain this better

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/full_model.py (hunk 5 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/full_model.py:53]: I don't expect them to be actively developed in the future, but moving it might break some code now. Let's touch base on this later

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py (hunk 6 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py:124]: Instead, I just moved these inside the function

File: packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py (hunk 7 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py:151]: Now defaults to 0.0

File: packages/greedy_infomax/tests/unit/test_gradient_block.py (hunk 8 above)
Comment (alexcuozzo-1) [reply packages/greedy_infomax/tests/unit/test_gradient_block.py:36]: Done

File: projects/greedy_infomax/experiments/__init__.py (hunk 9 above)
Comment (alexcuozzo-1) [reply projects/greedy_infomax/experiments/__init__.py:26]: Done

**Review by karangrewal — APPROVED** [review]
